cli/lib/gateway.js

```diff
@@ -1,104 +1,200 @@
 'use strict';
 const path = require('path');
-const assert = require('assert');
-const cluster = require('cluster');
 const os = require('os');
 const fs = require('fs');
+const net = require('net');
 const edgeconfig = require('microgateway-config');
 const gateway = require('microgateway-core');
+const reloadCluster = require('./reload-cluster');
+const JsonSocket = require('./json-socket');
 const configLocations = require('../../config/locations');
-const agentConfig = require('../../lib/agent-config');
+const isWin = /^win/.test(process.platform);
+const ipcPath = configLocations.getIPCFilePath();
 
 const Gateway = function () {
-}
+};
 
 module.exports = function () {
   return new Gateway();
-}
+};
+
+Gateway.prototype.start =  (options) => {
+
+  try {
+    fs.accessSync(ipcPath, fs.F_OK);
+    console.error('Edgemicro seems to be already running.');
+    console.error('If the server is not running, it might because of incorrect shutdown of the prevous start.');
+    console.error('Try removing ' + ipcPath + ' and start again');
+    process.exit(1);
+  } catch (e) {
+    // Socket does not exist
+    // so ignore and proceed
+  }
 
-Gateway.prototype.start = function start(options, cb) {
   const source = configLocations.getSourcePath(options.org, options.env);
   const cache = configLocations.getCachePath(options.org, options.env);
-  const keys = { key: options.key, secret: options.secret };
-  const args = { target: cache, keys: keys,pluginDir:options.pluginDir };
-  const that = this;
-
-  if (cluster.isMaster) {
-    edgeconfig.get({ source: source, keys: keys }, function (err, config) {
-      
-      if(options.port){
-        config.edgemicro.port = parseInt(options.port);
+
+  const keys = {key: options.key, secret: options.secret};
+  const args = {target: cache, keys: keys, pluginDir: options.pluginDir};
+
+  edgeconfig.get({source: source, keys: keys},  (err, config) => {
+    if (err) {
+      const exists = fs.existsSync(cache);
+      console.error("failed to retieve config from gateway. continuing, will try cached copy..");
+      console.error(err);
+      if (!exists) {
+        console.error('cache configuration ' + cache + ' does not exist. exiting.');
+        return;
+      } else {
+        console.log('using cached configuration from %s', cache);
+        config = edgeconfig.load({source: cache});
+        if (options.port) {
+          config.edgemicro.port = parseInt(options.port);
+        }
+      }
+    } else {
+      edgeconfig.save(config, cache);
+    }
+
+    gateway(config);
```

Comment:
Interesting. Ill remove it. But just curious to know what it was doing earlier ?